docs/operating-scylla/procedures/cluster-management/remove-node.rst

```diff
@@ -56,12 +56,14 @@ Removing an Unavailable Node
 If the node status is **Down Normal (DN)**, you should try to restore it. Once the node is up, use the ``nodetool decommission``
 command (see `Removing a Running Node`_) to remove it.
 
-If all attempts to restore the node have failed and the node is down, you can remove the node by running the ``nodetool removenode``
+If all attempts to restore the node have failed and the node is **permanently down**, you can remove the node by running the ``nodetool removenode``
 command providing the Host ID of the node you are removing. See :doc:`nodetool removenode </operating-scylla/nodetool-commands/removenode>` for details.
 
 .. warning::
     * Using ``nodetool removenode`` is a fallback procedure that should only be used when a node is permanently down and cannot
-      be recovered. 
+      be recovered.
+    * It’s essential to ensure the removed node will **never** come back to the cluster, which **might lead to data loss**. To prevent 
```

Comment:
How can the removed node rejoining the cluster lead to data loss?

On the contrary, it may lead to data resurrection, if it reintroduces stale data that was deleted in the node absence and the tombstones were purged as well.

Reply:
It could lead to data loss, the removed node reintroduces its tokens which changes the replica sets of some token ranges, causing us to query "wrong" replica sets and not see some rows.

E.g. suppose that some token range had replica sets A,B,C and there was a row on A,B but not on C (it was a successful quorum write). Removed node restarts and reintroduces its tokens, causing that token range's replica sets to become B, C, D. We now query this key with quorum and the query may use C,D, so we "lost" the row.

Reply:
With repair-based node operations D (and C) will get the latest data from B during removenode of A.
Otherwise we could experience data loss once removenode completes and before repair is run, regardless of A's resurrection.

Reply:
the data "loss" scenario Kamil is worried about is the exposure from not running repair, so it involves yet another breach of procedures. Therefore I don't think it's worth mentioning in this context.
The real fix to these problems is finishing the automation of topology changes using raft so it will be much harder, if not virtually impossible, for users to shoot themselves in the foot.

Reply:
Also, @gleb-cloudius, during removenode, are we effectively increasing the quorum size so to prevent this data loss scenario like we so for decommission? removenode is abortable and the removed node can leaglly come back to life if the remove operation is aborted.

Reply:
**About data loss:**

> the data "loss" scenario Kamil is worried about is the exposure from not running repair, so it involves yet another breach of procedures. Therefore I don't think it's worth mentioning in this context.

If that's the case, I agree that we can skip the "data loss" part.

**About Raft:**

I understand that Raft solves the problem, but for now:
- Raft is not mandatory. 
- It's not even the default, so the user needs to enable it explicitly. 
- Raft schema changes have been implemented, but I don't think Raft topology changes are.
- In addition, the customer problem that started this conversation was reported for Scylla Enterprise, where Raft is not an option.

We need to help the users who cannot or don't want to enable Raft. I could add a comment that the note should be updated and later removed in future versions.

```
.. REMOVE IN FUTURE VERSIONS - Update the note to suggest enabling Raft when Raft topology changes are implemented. Remove the note below when Raft topology changes are mandatory.

It’s essential to ensure the removed node will **never** come back to the cluster, which might lead to **data resurrection.** To prevent the removed node from rejoining the cluster, remove that node from the cluster network or VPC.
```


Does that make sense from your perspective?

-------------------------------------------------------
Note:
The note says "the removed node", which assumes that the operation was not aborted and the node was successfully removed.

Reply:
> With repair-based node operations D (and C) will get the latest data from B during removenode of A.
Otherwise we could experience data loss once removenode completes and before repair is run, regardless of A's resurrection.

@bhalevy I didn't make the scenario precise enough

consider a key whose replicas are B,C,D before removenode
we removenode D, its replicas become A,B,C.
The user could run full repair at this point and everything, the cluster is perfectly in sync, doesn't matter.

Now, **after removenode and repair**, we perform a successful quorum write that ends up at A,B but not on C.
Then D restarts and reintroduces its tokens. The key's replica set becomes B,C,D again.
The user makes a quorum read and reads from C,D. The written row does not appear. We "lost" the row.

Therefore I still claim that a removed node coming back to life and rejoining the cluster can lead to data **loss**.